external_search.py

```diff
@@ -882,8 +574,738 @@ def create_new_version(cls, search_client, base_index_name, version=None):
+    # lower if there don't appear to be any misspelled words in the
+    # query string.
+    FUZZY_CIRCUIT_BREAKER = re.compile(
+        r'\b(%s)\b' % "|".join(FUZZY_CONFOUNDERS), re.I
+    )
+
+    def __init__(self, query_string, filter=None):
+        """Store a query string and filter.
+
+        :param query_string: A user typed this string into a search box.
+        :param filter: A Filter object representing the circumstances
+            of the search -- for example, maybe we are searching within
+            a specific lane.
+        """
+        self.query_string = query_string
+        self.filter = filter
+
+    def build(self):
+        """Make an Elasticsearch-DSL query object out of this query."""
+        query = self.query()
+
+        # Add the filter, if there is one.
+        if self.filter:
+            built_filter = self.filter.build()
+            if built_filter:
+                query = Q("filtered", query=query, filter=built_filter)
+
+        # There you go!
+        return query
+
+    def query(self):
+        """Build an Elasticsearch Query object for this query string.
+        """
+        query_string = self.query_string
+
+        # The search query will create a dis_max query, which tests a
+        # number of hypotheses about what the query string might
+        # 'really' mean. For each book, the highest-rated hypothesis
+        # will be assumed to be true, and the highest-rated titles
+        # overall will become the search results.
+        hypotheses = []
+
+        # Here are the hypotheses:
+
+        # The query string might appear in one of the standard
+        # searchable fields.
+        simple = self.simple_query_string_query(query_string)
+        self._hypothesize(hypotheses, simple)
+
+        # The query string might be a close match against title,
+        # author, or series.
+        self._hypothesize(
+            hypotheses,
+            self.minimal_stemming_query(query_string),
+            100
+        )
+
+        # The query string might be an exact match for title or
+        # author. Such a match would be boosted quite a lot.
+        self._hypothesize(
+            hypotheses,
+            self._match_phrase("title.standard", query_string), 200
+        )
+        self._hypothesize(
+            hypotheses,
+            self._match_phrase("author.standard", query_string), 200
+        )
+
+        # The query string might be a fuzzy match against one of the
+        # standard searchable fields.
+        fuzzy = self.fuzzy_string_query(query_string)
+        self._hypothesize(hypotheses, fuzzy, 1)
+
+        # The query string might contain some specific field matches
+        # (e.g. a genre name or target age), with the remainder being
+        # the "real" query string.
+        with_field_matches = self._parsed_query_matches(query_string)
+        self._hypothesize(
+            hypotheses, with_field_matches, 20, all_must_match=True
```

Comment:
This weight of 20 was originally a weight of 200. Weighting this hypothesis the same as the _match_phrase hypotheses above causes "Gumby in Love" to show up before "Modern Romance" in a search for `modern romance`. Unfortunately, the way it's weighted now, "Modern Romance" shows up before "Gumby in Love" in a search for `romance`, which isn't a disaster but this wasn't originally the case. This is the one thing about this branch I haven't been able to resolve to my satisfaction -- I don't see any significant differences in the generated Elasticsearch documents.

Reply:
When I run the tests locally with the old code I get a couple failures, and one is that the `modern romance` test puts Gumby first. I think it's one of those things where tf-idf is computed within the shard so if they get indexed in different shards it doesn't work. It's possible that's happening to you as well. That's why you really need to test with a realistic index - these small tests are nice but don't tell you much.

But when I run it locally with your branch it finds 50 results - I wonder if it's using the wrong index.

Reply:
It looks like the way you're creating the Search doesn't tell it what index to use, so it's searching across every index I have.

Reply:
I've fixed the index problem.

My local Elasticsearch and the one set up by Travis both have `number_of_shards:1`, but I'm comparing the results here and they have different values for `.meta.shard`.  I've added this to the debug portion of the code so it'll show up on Travis if the same problem exists there.